tests/fuzz/fuzz_exceptions.py

```diff
@@ -0,0 +1,548 @@
+          (5,  self.make_loop),
+        ]
+        if self.try_nesting:
+            options.append((10, self.make_throw))
+        else:
+            # Only rarely emit throws outside of a try.
+            options.append((2, self.make_throw))
+        if self.loop_nesting:
+            options.append((10, self.make_branch))
+        return pick(options, cursor.get_num(), cursor)
+
+    def make_nothing(self, cursor):
+        return ''
+
+    def make_logging(self, cursor):
+        if cursor.has_more():
+            return f'puts("log(-{cursor.get_int()})");'
+        self.logging_index += 1
+        return f'puts("log({self.logging_index})");'
+
+    def make_throw(self, cursor):
+        return f'throw {cursor.get_int()};'
+
+    def make_try(self, cursor):
+        self.try_nesting += 1
+        body = indent(self.make_statement(cursor.get_array()))
+        self.try_nesting -= 1
+        catch = indent(self.make_statement(cursor.get_array()))
+
+        return '''\
+try {
+%(body)s
+} catch(...) {
+%(catch)s
+}
+''' % locals()
+
+    def make_if(self, cursor):
+        if_arm = indent(self.make_statement(cursor.get_array()))
+
+        else_ = ''
+        if cursor.get_num() >= 0.5:
+            else_arm = indent(self.make_statement(cursor.get_array()))
+            else_ = '''\
+ else {
+%(else_arm)s
+}''' % locals()
+
+        return '''\
+if (getBoolean()) {
+%(if_arm)s
+}%(else_)s
+''' % locals()
+
+    def make_loop(self, cursor):
+        self.loop_nesting += 1
+        body = indent(self.make_statement(cursor.get_array()))
+        self.loop_nesting -= 1
+
+        return '''\
+while (getBoolean()) {
+%(body)s
+}
+''' % locals()
+
+    def make_branch(self, cursor):
+        assert self.loop_nesting
+        if cursor.get_num() < 0.5:
+            return 'break;'
+        else:
+            return 'continue;'
+
+    def get_types(self, node):
+        return [self.get_type(x) for x in arrayify(node)]
+
+    def get_type(self, node):
+        if numify(node) < 0.5:
+            return 'uint32_t'
+        return 'double'
+
+
+# Main harness
+
+
+def check_testcase(data, silent=True):
+    # Generate C++
+    CppTranslator(data).write(main='a.cpp', support='b.cpp')
+
+    # Compile with emcc, looking for a compilation error.
+    try:
+        subprocess.check_call(['./em++', 'a.cpp', 'b.cpp', '-sWASM_BIGINT',
+                               '-fwasm-exceptions'],
+                               stderr=subprocess.PIPE if silent else None)
+    except KeyboardInterrupt:
+        print('[stopping by user request]')
+        sys.exit(0)
+    except Exception:
+        return False
+
+    return True
+
+
+def reduce(data):
+    '''
+    Given a failing testcase, reduce the input data to create a reduced C++
+    testcase.
+    '''
+
+    assert not check_testcase(data)
+
+    # The input is structured. The simplest thing is to reduce on it in text
+    # form, so that we do not need to have references to nested things etc.
```

Comment:
I would expect it to be much simpler to reduce the nested structure directly than to reduce the text, but I might be wrong. Whenever you remove a subtree, you just have to keep it around to put back in case the reduction is unsuccessful. If you assign some node in the tree to `curr`, then modifications to `curr` will be visible from the top-level structure.